### PR TITLE
Hooks: Type package

### PR DESCRIPTION
**Pull request by sirreal**
## Description
Type the Hooks package and include TypeScript type declarations in the published package.

Part of #18838 

## How has this been tested?

Tests pass and types are generated without issue.

## Types of changes
New feature: Publish type declarations for the Hooks package.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by sirreal** [comment]
I'll fill in a description next week, I was excited the tests are passing because they've been unreliable 😅

File: packages/hooks/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+	"extends": "../../tsconfig.base.json",
+	"compilerOptions": {
+		"rootDir": "src",
+		"declarationDir": "build-types",
+		"types": [ "gutenberg-env" ]
+	},
+	"references": [],
```
Comment (sirreal) [inline packages/hooks/tsconfig.json:8]:
```suggestion
```

File: packages/hooks/src/createAddHook.js [hunk 2]
```diff
@@ -70,7 +74,7 @@ function createAddHook( hooks ) {
 			// we're adding would come after the current callback, there's no
 			// problem; otherwise we need to increase the execution index of
 			// any other runs by 1 to account for the added element.
-			( hooks.__current || [] ).forEach( ( hookInfo ) => {
+			hooks.__current.forEach( ( hookInfo ) => {
```
Comment (saramarcondes) [inline packages/hooks/src/createAddHook.js:77]: Why this change?

File: packages/hooks/src/createRemoveHook.js [hunk 3]
```diff
@@ -58,7 +63,7 @@ function createRemoveHook( hooks, removeAll ) {
 					// comes after the current callback, there's no problem;
 					// otherwise we need to decrease the execution index of any
 					// other runs by 1 to account for the removed element.
-					( hooks.__current || [] ).forEach( ( hookInfo ) => {
+					hooks.__current.forEach( ( hookInfo ) => {
```
Comment (saramarcondes) [inline packages/hooks/src/createRemoveHook.js:66]: Same here. Is this totally safe?

File: packages/hooks/src/createAddHook.js (hunk 2 above)
Comment (sirreal) [reply packages/hooks/src/createAddHook.js:77]: Thank you for highlighting. I didn't investigate why this included the fallback but the types suggest it _is_ safe. Not that `createAddHook( hooks )` accepts a parameter of type `Hooks`:

https://github.com/WordPress/gutenberg/blob/77f872a1a50f39b8f618db147ac8a54bb3fef63e/packages/hooks/src/createAddHook.js#L19-L26

Which is an object with a `__current` property of type `Current[]`:

https://github.com/WordPress/gutenberg/blob/77f872a1a50f39b8f618db147ac8a54bb3fef63e/packages/hooks/src/index.js#L22-L30

We can also see that this package exposes `createHooks` and the default hooks:

https://github.com/WordPress/gutenberg/blob/77f872a1a50f39b8f618db147ac8a54bb3fef63e/packages/hooks/src/index.js#L53-L73

Inside `createHooks` —the only place these individual hook creators can be called as far as I can tell— `__current` does seem to be defined:

https://github.com/WordPress/gutenberg/blob/77f872a1a50f39b8f618db147ac8a54bb3fef63e/packages/hooks/src/createHooks.js#L15-L24

I do believe it's safe to remove the fallback.

**Review by sarayourfriend — APPROVED** [review]

File: packages/hooks/src/createAddHook.js (hunk 2 above)
Comment (saramarcondes) [reply packages/hooks/src/createAddHook.js:77]: Right on, thanks for digging into it.
